File: backend/app/schemas/development.py
import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel


# --- Career Pathways ---

class CareerPathwayCreate(BaseModel):
    name: str
    description: str | None = None
    levels: list[Any] | None = None


class CareerPathwayUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    levels: list[Any] | None = None


class CareerPathwayResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    name: str
    description: str | None = None
    levels: list[Any] | None = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# --- Development Milestones ---

class DevelopmentMilestoneCreate(BaseModel):
    title: str
    description: str | None = None
    target_date: date
    status: str = "pending"
    year: int
    sort_order: int = 0


class DevelopmentMilestoneUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    target_date: date | None = None
    completed_date: date | None = None
    status: str | None = None
    year: int | None = None
    sort_order: int | None = None


class DevelopmentMilestoneResponse(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID
    title: str
    description: str | None = None
    target_date: date
    completed_date: date | None = None
    status: str
    year: int
    sort_order: int
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# --- Development Checkpoints ---

class DevelopmentCheckpointCreate(BaseModel):
    checkpoint_date: date
    reviewer_id: uuid.UUID | None = None
    notes: str | None = None
    overall_assessment: str
    actions: list[Any] | None = None


class DevelopmentCheckpointUpdate(BaseModel):
    checkpoint_date: date | None = None
    reviewer_id: uuid.UUID | None = None
    notes: str | None = None
    overall_assessment: str | None = None
    actions: list[Any] | None = None


class DevelopmentCheckpointResponse(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID
    checkpoint_date: date
    reviewer_id: uuid.UUID | None = None
    notes: str | None = None
    overall_assessment: str
    actions: list[Any] | None = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# --- Development Goals ---

class DevelopmentGoalCreate(BaseModel):
    title: str
    description: str | None = None
    goal_type: str | None = None
    linked_competency_id: uuid.UUID | None = None
    target_date: date | None = None
    cost_estimate: float | None = None
    priority: str | None = None
    year: int | None = None
    linked_objective_id: uuid.UUID | None = None


class DevelopmentGoalUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: str | None = None
    evidence: str | None = None
    cost_estimate: float | None = None
    priority: str | None = None
    progress: int | None = None
    year: int | None = None
    actual_cost: float | None = None
    completed_date: date | None = None


class DevelopmentGoalResponse(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID
    title: str
    description: str | None = None
    goal_type: str | None = None
    linked_competency_id: uuid.UUID | None = None
    target_date: date | None = None
    status: str
    evidence: str | None = None
    cost_estimate: float | None = None
    priority: str | None = None
    progress: int = 0
    year: int | None = None
    linked_objective_id: uuid.UUID | None = None
    actual_cost: float | None = None
    completed_date: date | None = None
    created_at: datetime
    model_config = {"from_attributes": True}


# --- Development Plans ---

class DevelopmentPlanCreate(BaseModel):
    user_id: uuid.UUID
    review_period_id: uuid.UUID | None = None
    career_aspiration: str | None = None
    horizon_years: int | None = None
    start_date: date | None = None
    end_date: date | None = None
    career_pathway_id: uuid.UUID | None = None
    total_budget: float | None = None
    goals: list[DevelopmentGoalCreate] = []
    milestones: list[DevelopmentMilestoneCreate] = []


class DevelopmentPlanUpdate(BaseModel):
    status: str | None = None
    career_aspiration: str | None = None
    horizon_years: int | None = None
    start_date: date | None = None
    end_date: date | None = None
    career_pathway_id: uuid.UUID | None = None
    total_budget: float | None = None
    overall_progress: int | None = None


class DevelopmentPlanResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID
    review_period_id: uuid.UUID | None = None
    status: str
    career_aspiration: str | None = None
    horizon_years: int | None = None
    start_date: date | None = None
    end_date: date | None = None
    career_pathway_id: uuid.UUID | None = None
    overall_progress: int = 0
    total_budget: float | None = None
    goals: list[DevelopmentGoalResponse] = []
    milestones: list[DevelopmentMilestoneResponse] = []
    checkpoints: list[DevelopmentCheckpointResponse] = []
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}

File: frontend/src/api/development.ts
import { api } from './client';

export interface DevelopmentGoal {
  id: string;
  plan_id: string;
  title: string;
  description: string | null;
  goal_type: string;
  linked_competency_id: string | null;
  target_date: string | null;
  status: string;
  evidence: string | null;
  cost_estimate: number | null;
  priority: string | null;
  progress: number;
  year: number | null;
  linked_objective_id: string | null;
  actual_cost: number | null;
  completed_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface DevelopmentMilestone {
  id: string;
  plan_id: string;
  title: string;
  description: string | null;
  target_date: string;
  completed_date: string | null;
  status: string;
  year: number;
  sort_order: number;
  created_at: string;
}

export interface DevelopmentCheckpoint {
  id: string;
  plan_id: string;
  checkpoint_date: string;
  reviewer_id: string | null;
  notes: string | null;
  overall_assessment: string;
  actions: unknown[] | null;
  created_at: string;
}

export interface CareerPathway {
  id: string;
  workspace_id: string;
  name: string;
  description: string | null;
  levels: Array<{ title: string; typical_years?: number; required_competencies?: string[] }> | null;
  created_at: string;
  updated_at: string;
}

export interface DevelopmentPlan {
  id: string;
  workspace_id: string;
  user_id: string;
  review_period_id: string | null;
  status: string;
  career_aspiration: string | null;
  horizon_years: number | null;
  start_date: string | null;
  end_date: string | null;
  career_pathway_id: string | null;
  overall_progress: number;
  total_budget: number | null;
  goals: DevelopmentGoal[];
  milestones: DevelopmentMilestone[];
  checkpoints: DevelopmentCheckpoint[];
  created_at: string;
  updated_at: string;
}

export interface SkillsGapItem {
  competency_id: string;
  competency_name: string;
  required_level: string | null;
  current_level: string | null;
  gap: boolean;
}

const base = (ws: string) => `/workspaces/${ws}/development`;

export const developmentApi = {
  list: (workspaceId: string, params?: { user_id?: string }) =>
    api.get<DevelopmentPlan[]>(base(workspaceId), { params }),

  create: (workspaceId: string, data: { user_id: string; review_period_id?: string; career_aspiration?: string; horizon_years?: number; start_date?: string; end_date?: string; career_pathway_id?: string; total_budget?: number; goals?: Partial<DevelopmentGoal>[]; milestones?: Partial<DevelopmentMilestone>[] }) =>
    api.post<DevelopmentPlan>(base(workspaceId), data),

  update: (workspaceId: string, planId: string, data: Partial<DevelopmentPlan>) =>
    api.put<DevelopmentPlan>(`${base(workspaceId)}/${planId}`, data),

  addGoal: (workspaceId: string, planId: string, data: Partial<DevelopmentGoal>) =>
    api.post<DevelopmentGoal>(`${base(workspaceId)}/${planId}/goals`, data),

  updateGoal: (workspaceId: string, planId: string, goalId: string, data: Partial<DevelopmentGoal>) =>
    api.put<DevelopmentGoal>(`${base(workspaceId)}/${planId}/goals/${goalId}`, data),

  // Milestones
  addMilestone: (workspaceId: string, planId: string, data: Partial<DevelopmentMilestone>) =>
    api.post<DevelopmentMilestone>(`${base(workspaceId)}/${planId}/milestones`, data),

  updateMilestone: (workspaceId: string, planId: string, milestoneId: string, data: Partial<DevelopmentMilestone>) =>
    api.put<DevelopmentMilestone>(`${base(workspaceId)}/${planId}/milestones/${milestoneId}`, data),

  deleteMilestone: (workspaceId: string, planId: string, milestoneId: string) =>
    api.delete(`${base(workspaceId)}/${planId}/milestones/${milestoneId}`),

  // Checkpoints
  listCheckpoints: (workspaceId: string, planId: string) =>
    api.get<DevelopmentCheckpoint[]>(`${base(workspaceId)}/${planId}/checkpoints`),

  addCheckpoint: (workspaceId: string, planId: string, data: Partial<DevelopmentCheckpoint>) =>
    api.post<DevelopmentCheckpoint>(`${base(workspaceId)}/${planId}/checkpoints`, data),

  updateCheckpoint: (workspaceId: string, planId: string, checkpointId: string, data: Partial<DevelopmentCheckpoint>) =>
    api.put<DevelopmentCheckpoint>(`${base(workspaceId)}/${planId}/checkpoints/${checkpointId}`, data),

  // Career Pathways
  listPathways: (workspaceId: string) =>
    api.get<CareerPathway[]>(`${base(workspaceId)}/pathways`),

  createPathway: (workspaceId: string, data: Partial<CareerPathway>) =>
    api.post<CareerPathway>(`${base(workspaceId)}/pathways`, data),

  updatePathway: (workspaceId: string, pathwayId: string, data: Partial<CareerPathway>) =>
    api.put<CareerPathway>(`${base(workspaceId)}/pathways/${pathwayId}`, data),

  deletePathway: (workspaceId: string, pathwayId: string) =>
    api.delete(`${base(workspaceId)}/pathways/${pathwayId}`),

  // Skills Gap
  skillsGap: (workspaceId: string, planId: string) =>
    api.get<SkillsGapItem[]>(`${base(workspaceId)}/${planId}/skills-gap`),

  // Timeline
  timeline: (workspaceId: string, params?: { user_id?: string }) =>
    api.get(`${base(workspaceId)}/timeline`, { params }),
};
